apps/keyboard/js/settings/word_list_converter.js

```diff
@@ -0,0 +1,498 @@
+
+  root = this.balanceTree(root);
+
+  return root;
+};
+
+
+
+var TSTBuilder = function(words) {
+  this.words = words;
+
+  this._built = false;
+
+  this.maxWordLength = 0;
+  this.characterFrequency = {};
+  this.tstRoot = null;
+};
+
+TSTBuilder.prototype.debug = function(msg) {
+  if (_DEBUG) {
+    console.log(msg);
+  }
+};
+
+TSTBuilder.prototype.build = function() {
+  var tstRoot = null;
+  var tree = new TSTTree();
+
+  this.words.forEach(function(word) {
+    // Find the longest word in the dictionary
+    this.maxWordLength = Math.max(this.maxWordLength, word.length);
+
+    tstRoot = tree.insert(tstRoot, word + _EndOfWord);
+
+    // keep track of the letter frequencies
+    word.split('').forEach(function(ch) {
+      if (ch in this.characterFrequency) {
+        this.characterFrequency[ch]++;
+      } else {
+        this.characterFrequency[ch] = 1;
+      }
+    }, this);
+  }, this);
+
+  tstRoot = tree.balance(tstRoot);
+
+  this.tstRoot = tstRoot;
+  this._built = true;
+};
+
+TSTBuilder.prototype.getTreeRoot = function() {
+  if (!this._built) {
+    throw Error('TST not built yet.');
+  }
+
+  return this.tstRoot;
+};
+
+TSTBuilder.prototype.getMaxWordLength = function() {
+  if (!this._built) {
+    throw Error('TST not built yet.');
+  }
+
+  return this.maxWordLength;
+};
+
+TSTBuilder.prototype.getCharacterFrequency = function() {
+  if (!this._built) {
+    throw Error('TST not built yet.');
+  }
+
+  return this.characterFrequency;
+};
+
+
+
+var TSTSerializer = function(tstRoot) {
+  this._tstRoot = tstRoot;
+  this._output = null;
+};
+
+TSTSerializer.prototype.debug = function(msg) {
+  if (_DEBUG) {
+    console.log(msg);
+  }
+};
+
+// Serialize the tree to an array. Do it depth first, folling the
+// center pointer first because that might give us better locality
+TSTSerializer.prototype._serializeNode = function(node) {
+  this._output.push(node);
+
+  // JSConv: original offset assignment is removed as
+  // offset will be overwritten anyway in computeOffsets.
+
+  if (node.ch == _EndOfWord && node.center) {
+    this.debug('nul node with a center!');
+  }
+  if (node.ch != _EndOfWord && !node.center) {
+    this.debug('char node with no center!');
+  }
+
+  // do the center node first so words are close together
+  if (node.center) {
+    this._serializeNode(node.center);
+  }
+
+  if (node.left) {
+    this._serializeNode(node.left);
+  }
+
+  if (node.right) {
+    this._serializeNode(node.right);
+  }
+};
+
+TSTSerializer.prototype.serializeToNodes = function() {
+  this._output = [];
+  this._serializeNode(this._tstRoot);
+
+  return this._output;
+};
+
+
+
+
+// JSConv:
+// In the JS version, since we're not directly writing to a file,
+// 'output' is a JS array. We convert to UInt8Array when we
+// finishes pushing to 'output'. This is because UInt8Array's length
+// has to be decided at instantiation.
+// XXX: See if we can pre-determine the length when instantiating
+//      the buffer.
```

Comment:
After I've now studied the blob format and generation code more thoroughly, I believe this pre-determination is quite possible. I'll do that in a follow-up bug.